include/exadg/fluid_structure_interaction_precice/precice_adapter.h

```diff
@@ -0,0 +1,371 @@
+                     const std::string &                                                 mesh_name,
+                     const std::vector<std::string> & read_data_name);
+
+  /**
+   * @brief      Advances preCICE after every timestep
+   */
+  void
+  advance(const double computed_timestep_length);
+
+  /**
+   * @brief      Saves current state of time dependent variables in case of an
+   *             implicit coupling
+   *
+   * @param[in]  state_variables Vector containing all variables to store as
+   *             reference
+   *
+   * @note       This function only makes sense, if it is used with
+   *             @p reload_old_state_if_required. Therefore, the order, in which the
+   *             variables are passed into the vector must be the same for
+   *             both functions.
+   */
+  void
+  save_current_state_if_required(const std::function<void()> & save_state);
+
+  /**
+   * @brief      Reloads the previously stored variables in case of an implicit
+   *             coupling. The current implementation supports subcycling,
+   *             i.e. previously refers o the last time
+   *             @p save_current_state_if_required() has been called.
+   *
+   * @param[out] state_variables Vector containing all variables to reload
+   *             as reference
+   *
+   * @note       This function only makes sense, if the state variables have been
+   *             stored by calling @p save_current_state_if_required. Therefore,
+   *             the order, in which the variables are passed into the
+   *             vector must be the same for both functions.
+   */
+  void
+  reload_old_state_if_required(const std::function<void()> & reload_old_state);
+
+
+  void
+  write_data(const std::string & write_mesh_name,
+             const std::string & write_data_name,
+             const VectorType &  write_data,
+             const double        computed_timestep_length);
+
+  std::vector<Tensor<1, dim>>
+  read_block_data(const std::string & mesh_name, const std::string & data_name) const;
+
+  /**
+   * @brief is_coupling_ongoing Calls the preCICE API function isCouplingOnGoing
+   *
+   * @return returns true if the coupling has not yet been finished
+   */
+  bool
+  is_coupling_ongoing() const;
+
+  /**
+   * @brief is_time_window_complete Calls the preCICE API function isTimeWindowComplete
+   *
+   * @return returns true if the coupling time window has been completed in the current
+   *         iteration
+   */
+  bool
+  is_time_window_complete() const;
+
+
+
+private:
+  // public precice solverinterface, needed in order to steer the time loop
+  // inside the solver.
+  std::shared_ptr<precice::SolverInterface> precice;
+  /// The objects handling reading and writing data
+  std::map<std::string, std::shared_ptr<CouplingInterface<dim, data_dim, VectorizedArrayType>>>
+    writer;
+  std::map<std::string, std::shared_ptr<CouplingInterface<dim, data_dim, VectorizedArrayType>>>
+    reader;
+
```

Comment:
Design-wise, the adapter class is supposed to provide a unified interface for the user so that only this class needs to be instantiated and configured in the user code. The class `CouplingInterface` is handles the different methods depending on what you want:

In case you like to use quadrature points for your coupling, the `QuadInterface` provides all functionality for reading and writing data on quadrature points. In case you like to use DoFs (global vectors) for your coupling, the `DoFInterface` provides all functionality for reading and writing from DoFs. There are mainly three reasons to put functionality for data reading and writing in the same class

1. they both require to define a coupling interface (i.e. pass coordinate information to preCICE) in the same way
2. the functionality for data reading and writing are usually very similar (e.g., extracting data from and writing data into a global vector)
3. we might use the same data locations (the same CouplingInterface instance) for reading and writing, which seems a bit more intricate in case we split reading and writing data into different classes

The main reason why it looks odd in this context is that ExaDG (the solver used here) has already coupling functionality and that's rather unusual. After I finished the coding work here, I even thought that it might have been easier to start from the IcNS and the solid solver example instead of the coupled variants used here.

In the coupled solver here, the coupling facilities of ExaDG are responsible for data reading and I removed the functionality from the `CouplingInterface` in order to reduce the amount of code in this PR. I only realized later on that the ExaDG interface class now 'only' overrides the boundary data in the `FunctionCache`.

I understand that it might look tempting for now to split the functionality according to reading and writing into different classes. However, merging the functionality of the ExaDG 'reader' and the Adapter 'writer' into a single class seems the more practical way to me. Do you agree?

Reply:
I had a closer look at `QuadInterface` vs. `DoFInterface`. I think it makes sense to realize these two classes not in the same class, in particular since the data structures `interface_node_ids` are not equivalent. However, would it make sense to implement the common parts of `define_coupling_mesh()` in a base class, while the derived classes have to implement a virtual function `initialize_interface_node_ids()` (in order to avoid code duplication between the two `define_coupling_mesh()` implementations for Quad/DoF).

Reply:
What is the purpose of `ExaDGInterface`? Are `using FEFaceIntegrator ...` and `using value_type ...` really necessary in this class?

I found that `ExaDGInterface` is used in `Adapter::add_read_interface()`, while `QuadInterface`, `DoFInterface` are used in `Adapter::add_write_interface()`. I still believe that this logic somehow needs to be reflected in the class names and/or the class hierarchy to make the code more understandable.

Reply:
> However, merging the functionality of the ExaDG 'reader' and the Adapter 'writer' into a single class seems the more practical way to me. Do you agree?

Do you mean that one could completely integrate `ExaDGInterface` into `Adapter`? Sound plausible.